src/diffusers/models/transformers/transformer_mochi.py

```diff
@@ -0,0 +1,390 @@
+        pos = self._get_positions(num_frames, height, width, device, dtype)
+        rope_cos, rope_sin = self._create_rope(pos_frequencies, pos)
+        return rope_cos, rope_sin
+
+
+@maybe_allow_in_graph
+class MochiTransformer3DModel(ModelMixin, ConfigMixin):
+    r"""
+    A Transformer model for video-like data introduced in [Mochi](https://huggingface.co/genmo/mochi-1-preview).
+
+    Args:
+        patch_size (`int`, defaults to `2`):
+            The size of the patches to use in the patch embedding layer.
+        num_attention_heads (`int`, defaults to `24`):
+            The number of heads to use for multi-head attention.
+        attention_head_dim (`int`, defaults to `128`):
+            The number of channels in each head.
+        num_layers (`int`, defaults to `48`):
+            The number of layers of Transformer blocks to use.
+        in_channels (`int`, defaults to `12`):
+            The number of channels in the input.
+        out_channels (`int`, *optional*, defaults to `None`):
+            The number of channels in the output.
+        qk_norm (`str`, defaults to `"rms_norm"`):
+            The normalization layer to use.
+        text_embed_dim (`int`, defaults to `4096`):
+            Input dimension of text embeddings from the text encoder.
+        time_embed_dim (`int`, defaults to `256`):
+            Output dimension of timestep embeddings.
+        activation_fn (`str`, defaults to `"swiglu"`):
+            Activation function to use in feed-forward.
+        max_sequence_length (`int`, defaults to `256`):
+            The maximum sequence length of text embeddings supported.
+    """
+
+    _supports_gradient_checkpointing = True
+
+    @register_to_config
+    def __init__(
+        self,
+        patch_size: int = 2,
+        num_attention_heads: int = 24,
+        attention_head_dim: int = 128,
+        num_layers: int = 48,
+        pooled_projection_dim: int = 1536,
+        in_channels: int = 12,
+        out_channels: Optional[int] = None,
+        qk_norm: str = "rms_norm",
+        text_embed_dim: int = 4096,
+        time_embed_dim: int = 256,
+        activation_fn: str = "swiglu",
+        max_sequence_length: int = 256,
+    ) -> None:
+        super().__init__()
+
+        inner_dim = num_attention_heads * attention_head_dim
+        out_channels = out_channels or in_channels
+
+        self.patch_embed = PatchEmbed(
+            patch_size=patch_size,
+            in_channels=in_channels,
+            embed_dim=inner_dim,
+            pos_embed_type=None,
+        )
+
+        self.time_embed = MochiCombinedTimestepCaptionEmbedding(
+            embedding_dim=inner_dim,
+            pooled_projection_dim=pooled_projection_dim,
+            text_embed_dim=text_embed_dim,
+            time_embed_dim=time_embed_dim,
+            num_attention_heads=8,
+        )
+
+        self.pos_frequencies = nn.Parameter(torch.empty(3, num_attention_heads, attention_head_dim // 2))
```

Comment:
cc @yiyixuxu Mochi uses learned RoPE frequencies. Ideally we'd like to always be in fp32 even when the module is converted to lower dtype for inference. The downcasting is handled in apply_rotary_emb in `MochiAttentionProcessor2_0`

Reply:
@sayakpaul I've used the `_keep_modules_in_fp32` based on Dhruv's suggestion. Just to confirm, just adding `_keep_modules_in_fp32 = ["pos_frequencies"]` is the intended way of using it, yes? If not, could you let me know what is?

Reply:
you can test it out! 
after you load the transformer in bfloat16, you can check the dtype of all the layers, all should be in bfloat16 except for this one

Reply:
This won't work but the following will:

```python
from diffusers import MochiTransformer3DModel 
from accelerate import init_empty_weights
import torch 

with torch.no_grad() and init_empty_weights():
    model = MochiTransformer3DModel.from_config(
        "diffusers-internal-dev/mochi-1-preview-diffusers", subfolder="transformer", torch_dtype=torch.float16
    )
    print(model.pos_frequencies.dtype)
```

Why? The reason is that we leverage `load_checkpoint_and_dispatch()` from `accelerate` to load a sharded checkpoint:
https://github.com/huggingface/diffusers/blob/298ab6eb01f3ef475c15218ea87de1494e1250aa/src/diffusers/models/modeling_utils.py#L886

When we do that, we cannot bake in specific parameter conversion types like above. The above works because it's empty state dict. 

The following test works
https://github.com/huggingface/diffusers/blob/298ab6eb01f3ef475c15218ea87de1494e1250aa/tests/quantization/bnb/test_mixed_int8.py#L181

Because we merge checkpoints when using quantization (which is already known). 

This will be resolved after @SunMarc's PR on eliminating `load_checkpoint_and_dispatch()` and making it similar to how it's done in `transformers`. Will work with Marc on this.